### ISSUE TITLE
`consensus_decode` should take `&mut D`?

### ISSUE THREAD

**Issue by dpc** [rust-bitcoin/rust-bitcoin#1020]
After taking a stab at #1019 , it seems to me that `consensus_decode` should take `d: &mut D` and not as it currently does `d: D`. The reason is - it is basically a wrapper around [`Read::read`](https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read) which takes `&mut self`.

Instead of passing `&mut d` when descending to sub-decoding, the code could just pass `d` and it would work. Internally it would avoid creating `&mut &mut &mut &mut T` which seems kind of undesirable, even if compiler probably can get rid of these needless references.

**Comment by apoelstra** [comment]
I could be convinced. I dislake `&mut D` because it's a little ugly/unergonomic and strictly speaking it's less general since `&mut D` is `Read` wheneve `D` is.

But you may be correct that the tendency to create giant `&mut &mut ... &mut` chains is even worse. In rust-miniscript we were forced to use `&mut D` in a lot of places where we had recursive calls, because otherwise the compiler would generate an infinite `&mut` chain.

In summary, I have no strong opinions here.

**Comment by dpc** [comment]
>  it's less general since &mut D is Read wheneve D is

Indeed `d: D` is more general that `d : &mut D`. But since the trait implementations are always based on `Read::read`, I would argue that it's too-general, because what we really need is `&mut Read` in the first place.

**Comment by dpc** [comment]
Actually, no. Scratch that. Wrong way to think about.

We need to think separately about trait bond and function.

A trait bond `D: Read` is most general, yes. And we will keep it.

However

```rust
fn foo_ref(f: &mut Foo);
```

is more general **from the caller perspective**.
than 

```rust
fn foo_owned(f: Foo);
```

It is sometimes impossible to use (call) `foo_owned`, while being able to call `foo_ref`.  And it is reversed from the implementation perspective: It is sometimes impossible to **implement** `foo_ref`, while it might be possible to implement `foo_owned`. This is a tradeoff between freedom/flexibility between caller and implementation.


Because of the current too strict calling convention, we are sometimes unable to **call** `D::consensus_decode` because we need to give up ownership, while we we want to use it again. Also, not that our escape hatch is calling with `&mut d`, which is changing the type of the trait to `&mut D`.

The above point might be very important, because even in the compiler can optimize the runtime cost of `&mut &mut &mut D` it probably can no deduplicate monorphizations of all these needless types. So we end up with `impl Decodable for X` 
, `impl Decodable for &mut X` , `impl Decodable for &mut &mut X` , `impl Decodable for &mut &mut &mut X`, etc. **which is increasing compilation time and/or blowing up the binary size**.

**Comment by Kixunil** [comment]
I remember some Rust devs saying `d: D` is better but the arguments about monomorphisation sound very strong to me and I'm now in favor of `d: &mut D`. If someone wants to disprove it, bring it on!

**Comment by dpc** [comment]
Doesn't hurt to get a second opinion: https://users.rust-lang.org/t/taking-r-mut-r-vs-mut-r-r-where-r-io-read/76780
